src/dune_pkg/opam.ml

```diff
@@ -0,0 +1,188 @@
+             a subdirectory named \"packages\""
+            opam_repo_dir_path
+        ];
+    { packages_dir_path }
+
+  (* Return the path to an "opam" file describing a paticular package
+     (name and version) from this opam repository. *)
+  let get_opam_file_path t opam_package =
+    t.packages_dir_path
+    / OpamPackage.name_to_string opam_package
+    / OpamPackage.to_string opam_package
+    / "opam"
+
+  (* Reads an opam package definition from an "opam" file in this repository
+     corresponding to a package (name and version). *)
+  let load_opam_package t opam_package =
+    let opam_file_path = get_opam_file_path t opam_package in
+    if not (Sys.file_exists opam_file_path) then
+      User_error.raise
+        [ Pp.textf
+            "Couldn't find package file for \"%s\". It was expected to be \
+             located in \"%s\" but this file does not exist"
+            (OpamPackage.to_string opam_package)
+            opam_file_path
+        ];
+    OpamFile.OPAM.read (OpamFile.make (OpamFilename.raw opam_file_path))
+end
+
+module Env = struct
+  type t = OpamVariable.variable_contents OpamVariable.Map.t
+
+  let empty : t = OpamVariable.Map.empty
+
+  let global () : t =
+    OpamGlobalState.with_ `Lock_none (fun global_state ->
+        OpamVariable.Map.filter_map
+          (fun _variable (contents, _description) -> Lazy.force contents)
+          global_state.global_variables)
+
+  let find_by_name (t : t) ~name =
+    OpamVariable.Map.find_opt (OpamVariable.of_string name) t
+end
+
+(* A custom solver context based on [Opam_0install.Dir_context] with a set
+   of local packages (ie. the packages defined in the current project).
+   When looking up a package during solving, the local packages are
+   searched before falling back to packages defined in a directory in the
+   style of opam-repository. *)
+module Solver_context = struct
+  module Dir_context = Opam_0install.Dir_context
+
+  (* Version to use for local packages with no version number *)
+  let local_package_default_version = OpamPackage.Version.of_string "LOCAL"
+
+  type t =
+    { dir_context : Dir_context.t
+    ; local_packages : OpamFile.OPAM.t OpamPackage.Name.Map.t
+    }
+
+  type rejection = Dir_context.rejection
+
+  let pp_rejection = Dir_context.pp_rejection
+
+  let candidates t name =
+    match OpamPackage.Name.Map.find_opt name t.local_packages with
+    | None -> Dir_context.candidates t.dir_context name
+    | Some opam_file ->
+      let version =
+        Option.value opam_file.version ~default:local_package_default_version
+      in
+      [ (version, Ok opam_file) ]
+
+  let user_restrictions t = Dir_context.user_restrictions t.dir_context
+
+  let filter_deps t = Dir_context.filter_deps t.dir_context
+
+  let create ~env ~repo ~local_packages =
+    let env name = Env.find_by_name env ~name in
+    let { Repo.packages_dir_path } = repo in
+    let dir_context =
+      Dir_context.create ~prefer_oldest:true
```

Comment:
So we're going with MVS?

Reply:
Ah good question! MVS would be my preference but this merits further discussion. @rgrinberg thoughts?

Reply:
It's a good default, but we could make this configurable.

Reply:
I think it is an interesting choice for the prototype and solves a lot of "lower bounds" failures that we frequently have in OPAM, so I'm in favor of trying and collecting feedback.